ballerina/soap_utils.bal

```diff
@@ -1,106 +1,198 @@
+            if serverPublicKey !is () {
+                byte[] signatureData = check wssec:getSignatureData(soapEnvelope);
+                boolean verify = check wssec:verifyData((soapEnvelope/<soap:Body>/*).toString().toBytes(),
+                                                        signatureData, serverPublicKey, signatureAlgorithm);
+                if !verify {
+                    return error Error("Signature verification of the SOAP envelope has been failed");
+                }
+            }
+        }
+        return soapEnvelope;
+    } on fail var e {
+        return error Error("Outbound security configurations do not match with the SOAP response. ", e.cause());
+    }
+}
+
+string path = "";
+
+public function sendReceive(xml|mime:Entity[] body, http:Client httpClient, string? soapAction = (),
+                            map<string|string[]> headers = {}, boolean soap12 = true) returns xml|Error {
+    http:Request req;
+    if soap12 {
+        req = createSoap12HttpRequest(body, soapAction, headers);
+    } else {
+        req = createSoap11HttpRequest(body, <string>soapAction, headers);
+    }
+    do {
+        http:Response response = check httpClient->post(path, req);
+        if soap12 {
+            return check createSoap12Response(response);
+        }
+        return check createSoap11Response(response);
+    } on fail var err {
+        return error Error(SOAP_RESPONSE_ERROR, err);
+    }
+}
+
+public function sendOnly(xml|mime:Entity[] body, http:Client httpClient, string? soapAction = (),
+                         map<string|string[]> headers = {}, boolean soap12 = true) returns Error? {
+    http:Request req;
+    if soap12 {
+        req = createSoap12HttpRequest(body, soapAction, headers);
+    } else {
+        req = createSoap11HttpRequest(body, <string>soapAction, headers);
+    }
+    http:Response|http:ClientError response = httpClient->post(path, req);
+    if response is http:ClientError {
+        return error Error(response.message());
+    }
+}
+
+function createSoap11HttpRequest(xml|mime:Entity[] body,
+                                 string soapAction, map<string|string[]> headers = {}) returns http:Request {
+    http:Request req = new;
+    if body is xml {
+        req.setXmlPayload(body);
+    } else {
+        req.setBodyParts(body);
+    }
+    req.setHeader(mime:CONTENT_TYPE, mime:TEXT_XML);
+    req.addHeader(SOAP_ACTION, soapAction);
+    foreach string key in headers.keys() {
+        req.addHeader(key, headers[key].toBalString());
+    }
+    return req;
+}
+
+function createSoap12HttpRequest(xml|mime:Entity[] body, string? soapAction,
+                                 map<string|string[]> headers = {}) returns http:Request {
+    http:Request req = new;
+    if body is xml {
+        req.setXmlPayload(body);
+    } else {
+        req.setBodyParts(body);
+    }
+    if soapAction is string {
+        map<string> stringMap = {};
+        stringMap[ACTION] = "\"" + soapAction + "\"";

```

Comment:
can use a string template here

Reply:
Fixed in [e505be0](https://github.com/ballerina-platform/module-ballerina-soap/pull/103/commits/e505be0a3eea0e469582edc70458da1e1bfa61c2)